envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotsVersions.kt

```diff
@@ -28,18 +28,23 @@ internal class SnapshotsVersions {
                 null -> Version(clusters = ClustersVersion(newVersion()), endpoints = EndpointsVersion(newVersion()))
                 else -> Version(
                     clusters = selectClusters(previous, clusters),
-                    endpoints = selectEndpoints(previous, endpoints)
+                    endpoints = selectEndpoints(previous, endpoints, previous.clusters != clusters)
                 )
             }
             VersionsWithData(version, clusters, endpoints)
         }
         return versionsWithData!!.version
     }
 
+    /**
+     * When cluster change we should also send EDS.
+     */
     private fun selectEndpoints(
         previous: VersionsWithData,
-        endpoints: List<ClusterLoadAssignment>
-    ) = if (previous.endpoints == endpoints) previous.version.endpoints else EndpointsVersion(newVersion())
+        endpoints: List<ClusterLoadAssignment>,
+        clusterChanged: Boolean
+    ) = if (previous.endpoints == endpoints && !clusterChanged)
```

Comment:
Maybe switch to: ` !clusterChanged && previous.endpoints == endpoints` to not compare endpoints when clusters changed

Reply:
Done